src/common/heartbeatService.js

```diff
@@ -0,0 +1,105 @@
+const {
+  APPLICATION_TAG_KEY,
+  CLUSTER_TAG_KEY,
+  SERVICE_TAG_KEY,
+  SHARD_TAG_KEY,
+  COMPONENT_TAG_KEY,
+  HEART_BEAT_METRIC,
+  HEART_BEAT_INTERVAL,
+  NULL_TAG_VAL
+} = require('./constants');
+
+/**
+ * Service that periodically reports component heartbeats to Wavefront.
+ */
+class HeartbeatService {
+  /**
+   *
+   * @param {WavefrontDirectClient} wavefrontClient
+   * @param applicationTags
+   * @param {Array} components - Array of string indicates Components.
+   * @param source
+   */
+  constructor(wavefrontClient, applicationTags, components, source) {
+    this.wavefrontClient = wavefrontClient;
+    this.applicationTags = applicationTags;
+    this.source = source;
+    this.reportingIntervalSeconds = HEART_BEAT_INTERVAL;
+    this.heartbeatMetricTagsList = [];
+    this.customTagsSet = new Set();
+
+    if (components instanceof string) {
+      components = [components];
+    }
+
+    for (let component of components) {
+      let metricTags = new Map([
```

Comment:
generally speaking, Map is reserved for times when you intend to have non-string keys.  Most code ends up using objects as maps, since they're easier to work with.